File: test/domain/recovery/recovery_preview_test.dart
import 'package:buffer/domain/recovery/recovery_preview.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('RecoveryPreview.truncate', () {
    test(
      'given_input_exactly_80_chars_no_newline_when_truncate_then_unchanged_length_80',
      () {
        final input = 'a' * 80;
        final result = RecoveryPreview.truncate(input);
        expect(result, equals(input));
        expect(result.length, equals(80));
      },
    );

    test(
      'given_input_less_than_80_chars_no_newline_when_truncate_then_identity',
      () {
        const input = 'short string';
        final result = RecoveryPreview.truncate(input);
        expect(result, equals(input));
      },
    );

    test(
      'given_input_greater_than_80_chars_no_newline_when_truncate_then_first_80_chars_no_ellipsis',
      () {
        final input = 'a' * 120;
        final result = RecoveryPreview.truncate(input);
        expect(result.length, equals(80));
        expect(result, equals('a' * 80));
        expect(result.contains('…'), isFalse);
      },
    );

    test(
      'given_multiline_input_when_truncate_then_newlines_collapsed_length_leq_80_no_newline',
      () {
        // "line one\nline two\n" is 19 chars; then pad with 200 chars of 'x'
        final input = 'line one\nline two\n${'x' * 200}';
        final result = RecoveryPreview.truncate(input);
        expect(result.length, lessThanOrEqualTo(80));
        expect(result.contains('\n'), isFalse);
      },
    );

    test(
      'given_string_of_only_newline_runs_when_truncate_then_collapses_to_single_space',
      () {
        const input = '\n\n\n\n';
        final result = RecoveryPreview.truncate(input);
        expect(result, equals(' '));
        expect(result.length, equals(1));
        expect(result.contains('\n'), isFalse);
      },
    );

    test(
      'given_multiline_input_whose_collapsed_form_is_exactly_80_when_truncate_then_no_further_truncation',
      () {
        // Build a string whose collapsed form is exactly 80 chars.
        // "hello\nworld" collapsed = "hello world" (11 chars).
        // We need collapsed to be exactly 80: 39 + '\n' + 40 = "a*39 \n b*40" → collapsed = "a*39 b*40" = 80 chars.
        final input = '${'a' * 39}\n${'b' * 40}';
        final collapsed = '${'a' * 39} ${'b' * 40}'; // 39 + 1 + 40 = 80
        final result = RecoveryPreview.truncate(input);
        expect(result.length, equals(80));
        expect(result, equals(collapsed));
        expect(result.contains('\n'), isFalse);
      },
    );

    test(
      'given_emoji_straddling_position_79_to_80_when_truncate_then_length_leq_80_no_newline_no_surrogate_split',
      () {
        // An emoji like '😀' uses 2 UTF-16 code units (a surrogate pair).
        // Place 79 'a' chars then the emoji: total = 79 + 2 = 81 code units.
        // Truncation must NOT cut between the two surrogates of the emoji.
        final emoji = '\u{1F600}'; // 😀 — 2 UTF-16 code units
        expect(emoji.length, equals(2)); // verify it's a surrogate pair
        final input = '${'a' * 79}$emoji';
        expect(input.length, equals(81));

        // Must not throw a RangeError and output ≤ 80 code units.
        final result = RecoveryPreview.truncate(input);
        expect(result.length, lessThanOrEqualTo(80));
        expect(result.contains('\n'), isFalse);
        // Ensure no lone surrogate by checking we can encode the result.
        expect(() => result.codeUnits, returnsNormally);
      },
    );
  });
}

File: shared/src/commonTest/kotlin/com/paolosantucci/foglietto/shared/recovery/RecoveryPreviewTest.kt
package com.paolosantucci.foglietto.shared.recovery

import okio.Buffer
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

// Port of test/domain/recovery/recovery_preview_test.dart
// Oracle: Dart RecoveryPreview.truncate test suite (8 ported cases + 1 PORTING-ADDITION)
//
// Invariants of truncate:
//   - result.length <= 80 (UTF-16 code units)
//   - result contains no '\n'
//   - \s*\n+\s* runs collapsed to one space
//   - No surrogate pair is split
//   - No ellipsis appended
//   - Input <80 code units with no '\n' returned identity-equal

class RecoveryPreviewTest {

    @Test
    fun truncate_multilineInput_newlinesCollapsedToSpaces() {
        // "a\n\n  b\n c" -> "a b c"
        val result = RecoveryPreview.truncate("a\n\n  b\n c")
        assertEquals("a b c", result)
    }

    @Test
    fun truncate_onlyNewlines_collapsesToSingleSpace() {
        // "\n\n\n" -> " " (single space after collapse run — EC-09)
        val result = RecoveryPreview.truncate("\n\n\n")
        assertEquals(" ", result)
    }

    @Test
    fun truncate_emptyString_returnsEmpty() {
        // "" -> "" (EC-09)
        val result = RecoveryPreview.truncate("")
        assertEquals("", result)
    }

    @Test
    fun truncate_200CharsNoNewline_returnsExactly80Units() {
        // 200-char no-newline input -> exactly 80 UTF-16 code units, no ellipsis
        val input = "a".repeat(200)
        val result = RecoveryPreview.truncate(input)
        assertEquals(80, result.length)
        assertFalse(result.contains('…'))
        assertFalse(result.contains('\n'))
    }

    @Test
    fun truncate_exactly80CharsNoNewline_identityEqual() {
        // exactly-80 no-newline -> identity-equal (same content, length 80) (FR-05)
        val input = "a".repeat(80)
        val result = RecoveryPreview.truncate(input)
        assertEquals(input, result)
        assertEquals(80, result.length)
    }

    @Test
    fun truncate_79CharsNoNewline_returnedUnchanged() {
        // 79 chars no-newline -> unchanged, length 79 (FR-05)
        val input = "a".repeat(79)
        val result = RecoveryPreview.truncate(input)
        assertEquals(input, result)
        assertEquals(79, result.length)
    }

    @Test
    fun truncate_emojiCrossing80thBoundary_cutOnCodePointBoundary() {
        // Emoji (😀, U+1F600) is a UTF-16 surrogate pair (2 code units).
        // Place 79 'a' chars then the emoji: total = 79 + 2 = 81 code units.
        // The high surrogate falls at index 79 (0-based) = the 80th code unit.
        // Truncation must NOT split the surrogate pair; result <= 80 code units.
        val emoji = "😀" // 😀 as explicit surrogate pair
        assertEquals(2, emoji.length) // verify it IS a surrogate pair
        val input = "a".repeat(79) + emoji
        assertEquals(81, input.length)

        val result = RecoveryPreview.truncate(input)
        assertTrue(result.length <= 80)
        assertFalse(result.contains('\n'))
        // Verify no lone surrogate by checking that codeUnits are accessible
        // (any lone surrogate would indicate a split pair)
        val units = result.map { it.code }
        assertTrue(units.isNotEmpty() || result.isEmpty())
    }

    @Test
    fun maxLength_equals80() {
        // The constant MAX_LENGTH must equal 80 (named const val, not a magic number) (R-A2)
        assertEquals(80, RecoveryPreview.MAX_LENGTH)
    }

    @Test
    fun truncate_512ByteHeadEndingMidUtf8_doesNotThrow() {
        // [PORTING-ADDITION] A 512-byte head whose last bytes form an incomplete UTF-8 sequence
        // (e.g. the first byte of a 3-byte CJK character at byte 512) decoded via okio
        // Buffer.readUtf8() must NOT throw; the result is a valid Kotlin String.
        // (FR-06, NFR-04, EC-10 — no Dart oracle equivalent; Kotlin/Native parity requirement)
        //
        // We construct a byte sequence: 510 bytes of ASCII 'a' (0x61) + 2 bytes of
        // a 3-byte UTF-8 CJK start (0xE4 0xB8 would be the first two bytes of U+4E00 '一').
        // The result after lenient decode should be a valid String (U+FFFD substitution ok).
        val buf = Buffer()
        // 510 bytes of 'a'
        repeat(510) { buf.writeByte(0x61) }
        // First two bytes of a 3-byte UTF-8 sequence (U+4E00 = 0xE4 0xB8 0x00),
        // intentionally incomplete: 0xE4 0xB8 without the third byte
        buf.writeByte(0xE4)
        buf.writeByte(0xB8)
        // Total: 512 bytes

        // readUtf8() with lenient decoding — must not throw
        val head = buf.readUtf8()

        // truncate must not throw on this input
        val result = RecoveryPreview.truncate(head)

        // result is a valid Kotlin String (no surrogate split); length <= 80
        assertTrue(result.length <= RecoveryPreview.MAX_LENGTH)
    }
}
